### PR TITLE
[MAJOR SEMVER BREAK] Rework all APIs

### PR DESCRIPTION
**Pull request by OxleyS**
This PR is a large set of breaking changes intended to:
- Provide fixes to https://github.com/DiamondLovesYou/rust-webm/issues/13 and https://github.com/DiamondLovesYou/rust-webm/issues/14 without requiring locking or `Weak` pointers
- Unify the API surface and its error handling around more idiomatic Rust, mainly the use of `Result`
- Clean up redundant APIs, or ones that may unexpectedly panic on the user
- Provide documentation for the entire `lib` API surface

As fixing the first issue required major semver changes anyway, I figured I might as well take the opportunity to do all these other improvements.

That said, there are some things in this PR that I left for future expansion - mainly this concerns the new `mux::Error` type, which is marked non-exhaustive and only contains an `Unknown` variant for now. In the future, the `lib` crate could perform more validation up-front and give more specific errors, but I felt this PR was beefy enough as is.

The all-at-once diff is pretty huge and hard to follow. Instead, I recommend going commit-by-commit, as I did my best to break everything up into concrete, incremental steps:
1. Everything up to and including the `Unified the return types of FFI...` commit is breaking changes for the `sys` crate, while the `lib` crate API does not change. Most of these changes are in support of the `lib` changes, however. They change the C++ so that tracks are referenced by number and not by pointer, and so that success/error info is returned in a consistent form.
2. From there, I removed segment-modifying methods from `AudioTrack` and `VideoTrack` and put them directly on the segment instead. At the end of the day, `AudioTrack` and `VideoTrack` become glorified track numbers. This solves the above two issues and lets me remove the `Arc<Mutex<>>` wrapping of the segment.
3. After that was ergonomics improvements, adding some documentation along the way. I also ended up moving `Writer` and `Segment` into their own source files as a way of keeping myself organized. This didn't change the export paths of this crate, however, it's still `mux::Segment` and `mux::Writer`.

Along the way, I found a few memory leaks which have also been plugged.

I tried to set up tooling to ensure the C++ declarations and Rust declarations in `sys` matched up, but couldn't get it to work. It may be good to double-check my work here in particular.

### CODE REVIEW

**Comment by OxleyS** [comment]
On reflection, I feel I could have presented this better - if you'd rather I split this up into multiple, more manageable PRs, I can do that. This PR might still be useful though for showing the bigger picture of where some of these changes were going.

File: src/lib/mux/segment.rs [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    /// Adds a new video track to this segment, returning its track number.
+    ///
+    /// You may request a specific track number using the `track_num` parameter. If one is specified, and this method
+    /// succeeds, the returned track number is guaranteed to match the requested one. If a track with that number
+    /// already exists, however, this method will fail. Leave as `None` to allow an available number to be chosen for
+    /// you.
+    ///
+    /// This method will fail if called after the first frame has been written.
```
Comment (kornelski) [inline src/lib/mux/segment.rs:99]: If you're redesigning the api, maybe use a builder or typestate pattern to ensure metadata is set before the frames?

File: src/lib/mux/segment.rs [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+    }
+
+    /// Sets color information for the specified video track.
+    ///
+    /// This method will fail if called after the first frame has been written.
+    pub fn set_color(
+        &mut self,
+        track: VideoTrack,
+        bit_depth: u8,
+        subsampling: (bool, bool),
```
Comment (kornelski) [inline src/lib/mux/segment.rs:199]: This would be clearer as an enum

File: src/lib/mux/segment.rs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        full_range: bool,
+    ) -> Result<(), Error> {
+        let (sampling_horiz, sampling_vert) = subsampling;
+        fn to_int(b: bool) -> i32 {
```
Comment (kornelski) [inline src/lib/mux/segment.rs:203]: c_int::from does that

File: src/lib/mux/writer.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        Self::make_writer(dest, get_pos_fn::<T>, Some(set_pos_fn::<T>))
+    }
+}
+
+/// The trait used by [`Segment`] to actually write out WebM data. This is implemented
+/// by [`Writer`], which in most cases is what you actually want to use.
+///
+/// ## Safety
+/// See the documentation for [`MkvWriter::mkv_writer`].
+pub unsafe trait MkvWriter {
```
Comment (kornelski) [inline src/lib/mux/writer.rs:174]: Does this need to exist? The impl is pretty specific, so I don't expect it to be useful to implement on anything else.

File: src/lib/lib.rs [hunk 5]
```diff
@@ -1,193 +1,59 @@
...
-        }
-    }
+    use crate::ffi;
 
-    #[doc(hidden)]
-    pub trait MkvWriter {
-        fn mkv_writer(&self) -> ffi::mux::WriterMutPtr;
-    }
+    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+    pub struct VideoTrack(TrackNum);
```
Comment (kornelski) [inline src/lib/lib.rs:40]: Copy and Eq promise it's going to be just a number. Is that a guarantee you want to give?

File: src/lib/mux/segment.rs [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            }
+            _ => Err(Error::Unknown),
+        }
+    }
+
+    /// Sets the CodecPrivate data a frame to the track with the specified track number. If you have a
+    /// [`VideoTrackNum`] or [`AudioTrackNum`], you can call `as_track_number()` to get the underlying [`TrackNum`].
+    ///
+    /// This method will fail if called after the first frame has been written.
+    pub fn set_codec_private(&mut self, track_number: u64, data: &[u8]) -> Result<(), Error> {
```
Comment (kornelski) [inline src/lib/mux/segment.rs:176]: the track number u64 should at least be TrackNum alias. Maybe impl Into<TrackNum> or some other trait to allow video and audio tracks seamlessly?

File: src/lib/mux/segment.rs [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+    /// succeeds, the returned track number is guaranteed to match the requested one. If a track with that number
+    /// already exists, however, this method will fail. Leave as `None` to allow an available number to be chosen for
+    /// you.
+    ///
+    /// This method will fail if called after the first frame has been written.
+    pub fn add_video_track(
+        &mut self,
+        width: u32,
+        height: u32,
+        desired_track_num: Option<i32>,
```
Comment (kornelski) [inline src/lib/mux/segment.rs:104]: Type used for track number is inconsistent

File: src/lib/mux/segment.rs [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+    ///
+    /// You may request a specific track number using the `track_num` parameter. If one is specified, and this method
+    /// succeeds, the returned track number is guaranteed to match the requested one. If a track with that number
+    /// already exists, however, this method will fail. Leave as `None` to allow an available number to be chosen for
+    /// you.
+    ///
+    /// This method will fail if called after the first frame has been written.
+    pub fn add_audio_track(
+        &mut self,
+        sample_rate: i32,
```
Comment (kornelski) [inline src/lib/mux/segment.rs:142]: These should be unsigned

**Review by kornelski — COMMENTED** [review]
Thanks for the PR.

File: src/lib/mux/writer.rs (hunk 4 above)
Comment (OxleyS) [reply src/lib/mux/writer.rs:174]: I was wondering that myself. If you don't see a need for it, then I am perfectly okay with removing it.

File: src/lib/lib.rs (hunk 5 above)
Comment (OxleyS) [reply src/lib/lib.rs:40]: I couldn't think of anything else we might want to put in here, considering we can't store a pointer to any `libwebm` object. Perhaps some cached metadata like the codec, which would make `Copy` less attractive.

That said, there is some intrinsic value in `Copy` - if I didn't know this crate and I saw a `video_track.clone()` I would be worried that it was unintentionally creating a second track. Still, if we feel uncomfortable making this commitment I can remove the `Copy` for now; we can always add it later.

`Eq`, I think, doesn't require the track to be just-a-number.

File: src/lib/mux/segment.rs (hunk 6 above)
Comment (OxleyS) [reply src/lib/mux/segment.rs:176]: Oops, this was an oversight. I'll do the same thing I did for `add_frame`.

File: src/lib/mux/segment.rs (hunk 7 above)
Comment (OxleyS) [reply src/lib/mux/segment.rs:104]: One snag here is that the underlying `libwebm` method takes an `i32` for some reason, so at some point we have to convert (and fail if it doesn't fit). We do have an error type for that now, though.

File: src/lib/mux/segment.rs (hunk 1 above)
Comment (OxleyS) [reply src/lib/mux/segment.rs:99]: It would be great if we could do something like:
```rust
SegmentBuilder::new()
  .with_app_name("my_cool_app")
  .with_video_track(VideoTrackBuilder::new(...))
  .with_audio_track(AudioTrackBuilder::new(...))
  .build()
```
But we'd have to figure out how to get `VideoTrack`s and `AudioTrack`s back to the user out of that. It's especially hairy because we don't know the underlying track numbers until we build the relevant `libwebm` objects, unless we were to forgo libwebm's track number selection and do it ourselves.

What we could do is something like this (names bikesheddable):
```rust
// `build()` creates the relevant `libwebm` objects
let mut segment: InitializedSegment = SegmentBuilder::new().with_app_name("hi").build();

// This creates the `libwebm` object on call
let video = segment.add_video_track(VideoTrackBuilder::new().with_color_info(...)).unwrap();

// From here, you can only call `add_frame()` and `finalize()`
let started: StartedSegment = segment.start();
```

Alternatively, if we're okay with building `libwebm` objects on the fly in `SegmentBuilder`:
```rust
// Builds libwebm objects as you call them
let builder = SegmentBuilder::new().with_app_name("hi");
let (builder, video_track) = segment.add_video_track(...);

let mut segment = segment.build(); // Mostly does nothing, just for typestate
segment.add_frame(...);
```
The only issue is that we'd need to make sure `libwebm` doesn't cause any side effects if the builder is dropped without calling `build()`. What do you think?

File: src/lib/mux/segment.rs [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+}
+
+/// A builder for [`Segment`].
+///
+/// Once you have a [`Writer`], you can use this to specify the tracks and track parameters you want, then build a
+/// [`Segment`], allowing you to write frames.
+pub struct SegmentBuilder<W: Write> {
+    segment: OwnedSegmentPtr,
+    writer: Writer<W>,
+}
```
Comment (OxleyS) [inline src/lib/mux/segment.rs:46]: I opted to eagerly construct FFI objects in the builder. This lets us avoid having separate public types for FFI-not-initialized and FFI-initialized states.

`libwebm` doesn't write anything to the `Writer` until you start writing media frames, or call `Finalize()`, so I don't expect any surprise I/O if a `SegmentBuilder` is dropped without calling `build()`.

File: src/lib/mux/segment.rs [hunk 10]
```diff
@@ -0,0 +1,375 @@
...
+    }
+
+    /// Sets color information for the specified video track.
+    pub fn set_color(
+        self,
+        track: VideoTrack,
+        bit_depth: u8,
+        subsampling: ColorSubsampling,
+        color_range: ColorRange,
+    ) -> Result<Self, Error> {
```
Comment (OxleyS) [inline src/lib/mux/segment.rs:205]: I decided against having something like `VideoTrackBuilder` for stuff like this, and just putting it on the segment builder instead, to keep the number of public types down and reduce duplication for methods shared between video and audio (such as setting CodecPrivate data).

File: src/sys/lib.rs [hunk 11]
```diff
@@ -25,69 +42,79 @@ pub mod mux {
...
-        pub fn mux_set_color(segment: VideoTrackMutPtr, bits: c_int, sampling_horiz: c_int, sampling_vert: c_int, full_range: c_int) -> c_int;
+        pub fn initialize_segment(segment: SegmentMutPtr, writer: WriterMutPtr) -> ResultCode;
+        #[link_name = "mux_set_color"]
+        pub fn mux_set_color(
+            segment: SegmentMutPtr,
+            video_track_num: TrackNum,
+            bits: u64,
+            sampling_horiz: u64,
+            sampling_vert: u64,
+            color_range: u64,
```
Comment (OxleyS) [inline src/sys/lib.rs:75]: `libwebm` itself takes `uint64_t` for all these parameters (including the color range, where you might expect the enum), so I've done the same here.

**Comment by OxleyS** [comment]
I believe with these latest commits, I've addressed the points of feedback. The one thing I did not do was remove `Copy` on `VideoTrack` and `AudioTrack`, as I couldn't get the ergonomics to work right with `add_frame()` and friends:
1. The existing `impl Into<TrackNum>` would consume the track, requiring every call with a `VideoTrack`/`AudioTrack` to use `.clone()`.
2. `impl Track` would work for `VideoTrack`/`AudioTrack`, but not raw `TrackNum`, preventing the writing of code agnostic to whether it's working with video or audio.
3. `AsRef<TrackNum>` ran into similar issues, you can't pass a raw `TrackNum` or even a `&TrackNum` this way.

**Comment by kornelski** [comment]
I've published it. Thanks for the PR!
